### ISSUE TITLE
Merging data frames

### ISSUE THREAD

**Issue by lesliedb** [dmcglinn/quant_methods#6]
Dan,

How am I supposed to merge the original data frame into the smaller data frame I created with richness???

I have successfully created a data frame (treeRichness) two columns: plotID and richness. Both containing 935 unique values with this code:
```{r}
treeRichness <- NULL

# For loop used to count species base on plotID subsets, and complie the subsets 
for(i in unique(tree$plotID)) {
  tempSubset <- subset(tree, plotID == i, select = c(plotID, species))
  tempSubset$richness <- length(unique(tempSubset$species))
  treeRichness <- rbind(treeRichness, tempSubset)
}

# Remove species from data frame
treeRichness$species <- NULL

# Remove row duplicates
treeRichness <- treeRichness[!duplicated(treeRichness),]
```
However, when I try to merge my original tree data frame, which has 8971 unique values, with treeRichness I always end up with 8971 unique values. I've tried every possible way to write the merge, but I still get the same result.

```{r}
test1 <- merge(treeRichness, tree, all.x = T, all.y = F)
dim(test1) #Output [1] 8971 14
test2 <- merge(treeRichness, tree, all.x = F, all.y = T)
dim(test2) #Output [1] 8971 14
test3 <- merge(treeRichness, tree, all.x = T, all.y = T)
dim(test3) #Output [1] 8971 14
test4 <- merge(treeRichness, tree, all.x = F, all.y = F)
dim(test4) #Output [1] 8971 14
test5 <- merge(tree, treeRichness, all.x = T, all.y = F)
dim(test5) #Output [1] 8971 14
test6 <- merge(tree, treeRichness, all.x = F, all.y = T)
dim(test6) #Output [1] 8971 14
test7 <- merge(tree, treeRichness, all.x = T, all.y = T)
dim(test7) #Output [1] 8971 14
test8 <- merge(tree, treeRichness, all.x = F, all.y = F)
dim(test8) #Output [1] 8971 14
```
I noticed in the [explanation](https://stat.ethz.ch/R-manual/R-devel/library/base/html/merge.html) of the merge function, there is a incomparables argument, which i'm quite not sure how to use but think it may be what i'm looking for with the merge function.

**Comment by dmcglinn** [comment]
Hey @lesliedb thanks for the question. I like your loop but here is a bit more streamlined version to accomplish the same thing without having to trim out the duplicates:
```
treeRichness <- NULL
uni_plotID <- unique(tree$plotID)
for(i in seq_along(uni_plotID)) {
    richness <- length(unique(tree$species[tree$plotID == uni_plotID[i]]))
    tempDat <- data.frame(plotID = uni_plotID[i], richness)
    treeRichness <- rbind(treeRichness, tempDat)
}
head(treeRichness)
dim(treeRichness)
```
I really appreciate your question about the merge. I thought this was a simple one that I had answered for @frie-knuth in #4 but it appears that I was mistaken in how the `merge` function works. I rechecked my code and I also see that when merging it expands the richness data.frame back to the 8971 rows. Here a completely reproducible example of how to accomplish the creation of richness, merge it with enviornment, and then aggregate it down to just the unique plotIDs
```
trees = read.csv('./data//treedata.csv')
sr = tapply(trees$spcode, trees$plotID, function(x) length(unique(x)))
summary(sr)
names(sr)[1:5]
sr_dat = data.frame(plotID = names(sr), sr=sr)
cols = c('plotID', 'utme', 'utmn', 'elev', 'tci', 'streamdist', 'disturb', 'beers')
sr_env = merge(sr_dat, trees[ , cols])
dim(sr_env)
# now aggregate sr_env by only taking the first value associated with each column for a given plotID
sr_env = aggregate(sr_env, by = list(sr_env$plotID),
                  FUN = function(x) x[1])
dim(sr_env)
head(sr_env)
```

**Comment by friedrichknuth** [comment]
tl;dr : Is there a way to accommodate and array or list of values for a single plotID as an explanatory variable input in a glm()?


*"...#now aggregate sr_env by only taking the first value associated with each column for a given plotID"*

It seems you are arbitrarily reducing the dataset here by assigning what happens to be the first value for a given environmental variable associated with a plotID. That was not exactly specified in the assignment and is what was causing the issue expressed in the addendum of my submission. 

I had circumvented the merge issue by combining the answer from Q2 and Q4, which results in a 935 line data.frame, as follows:

```{r}
trees = read.csv('./treedata.csv')

richness = with(trees, tapply(species, plotID, FUN = function(x) length(unique(x))))
rich_dat = data.frame(plotID = names(richness), plot_rich = richness)

utme = with(trees, tapply(utme, plotID, FUN = function(x) unique(x)))
utme_dat = data.frame(plotID = names(utme), plot_utme = utme)

utmn = with(trees, tapply(utmn, plotID, FUN = function(x) unique(x)))
utmn_dat = data.frame(plotID = names(utmn), plot_utmn = utmn)

elev = with(trees, tapply(elev, plotID, FUN = function(x) unique(x)))
elev_dat = data.frame(plotID = names(elev), plot_elev = elev)
    
tci = with(trees, tapply(tci, plotID, FUN = function(x) unique(x)))
tci_dat = data.frame(plotID = names(tci), plot_tci = tci)

streamdist = with(trees, tapply(streamdist, plotID, FUN = function(x) unique(x)))
streamdist_dat = data.frame(plotID = names(streamdist), plot_streamdist = streamdist)

disturb = with(trees, tapply(disturb, plotID, FUN = function(x) unique(x)))
disturb_dat = data.frame(plotID = names(disturb), plot_disturb = disturb)

beers = with(trees, tapply(beers, plotID, FUN = function(x) unique(x)))
beers_dat = data.frame(plotID = names(beers), plot_beers = beers)
                       
plot_env1 = merge(utme_dat, utmn_dat, by="plotID", all=TRUE)
plot_env2 = merge(plot_env1, elev_dat, by="plotID", all=TRUE)
plot_env3 = merge(plot_env2, tci_dat, by="plotID", all=TRUE)
plot_env4 = merge(plot_env3, streamdist_dat, by="plotID", all=TRUE)
plot_env5 = merge(plot_env4, disturb_dat, by="plotID", all=TRUE)
plot_env = merge(plot_env5, beers_dat, by="plotID", all=TRUE)

trees_wa = merge(rich_dat, plot_env, by="plotID", all=TRUE)
```
If you compare `trees_wa` with `sr_env` from your output, they both have 935 rows and the appropriate columns, except that `trees_wa` contains arrays (or lists) with all environmental information for a given column associated with a plotID, while `sr_env` contains integers and numeric values representing only the first environmental variable entry in a column for a given plotID. See `sapply(sr_env, class)` and `sapply(trees_wa, class)`. I understand that `sr_env` is what we want to go with and is the only valid input for `glm()` but merely taking the first entry for a given column by plotID seems like a subjective data reduction. 

Is there a way to accommodate and array or list of values for a single plotID as an explanatory variable input in a glm()?
